src/app/campaign/store/actions/campaign-action.creator.ts

```diff
@@ -5,6 +5,36 @@ import { HalDoc } from 'ngx-prx-styleguide';
 
 export class CampaignNew implements Action {
   readonly type = ActionTypes.CAMPAIGN_NEW;
+
+  constructor(public payload: {}) {}
+}
+
+export class CampaignDupFromForm implements Action {
+  readonly type = ActionTypes.CAMPAIGN_DUP_FROM_FORM;
+
+  constructor(public payload: { campaign: Campaign; flights: Flight[]; timestamp?: number }) {
```

Comment:
Following redux principles, the timestamp (nondeterministic) is created outside the reducer and the reducer uses it to generate the temporary flight ids keeping the reducer pure.

Reply:
😬 so it's actually the opposite, the tests provide a timestamp so they can validate against it for the flightIds, and the application expects the action to generate it. With how I've been doing things, I probably would have just provided it as part of the original payload object in the application, but I didn't because tbh I made this decision about where to actually put this code last. And I do kind of like this. I had forgotten about the idea of using action creators to generate this type of nondeterministic data. I've used it before, and it just kind of clicked again when working on this. So yeah, this is introducing a new idea to the application, but from my recollection, it is a standard practice.

Reply:
Ohh... okay.  Yeah, I follow.  I do that kind of thing all the time for testing (inject the timestamp, but default to Time.now()).

Reply:
I am really not proud of this but one day I found myself in a spot and figured out how to mock Date.now in specs https://github.com/PRX/grove.prx.org/blob/master/src/app/campaign/store/effects/campaign.effects.spec.ts#L258

Reply:
Ha!  Mocking Date.now is like a drug.  Now you won't be able to stop.